### PR TITLE
Euclidean Distance Transform

### PR DESCRIPTION
**Pull request by wkearn**
I needed a Euclidean Distance Transform for something I'm working on, so I wrote up the algorithm from [Maurer et al. 2003](http://ieeexplore.ieee.org/xpl/login.jsp?tp=&arnumber=1177156&url=http%3A%2F%2Fieeexplore.ieee.org%2Fstamp%2Fstamp.jsp%3Ftp%3D%26arnumber%3D1177156) Matlab uses for its `bwdist` function (but I wrote this from scratch after puzzling out the Maurer et al. pseudocode). 

This solves #65

Calling `(F,D) = bwdist(I)` where `I` is some binary image returns two results. F is the feature transform (i.e. the closest feature pixel given as a vector of its indices in I). D is the Euclidean Distance Transform calculated from F by computing the Euclidean distance from the pixel of interest to its feature pixel.

I'm fairly confident that it works on 2D images, but it should also work on multi-dimensional ones as well. The rudimentary tests that I've done on images of various dimensions seem to hold up. Let me know if you encounter any cases where it doesn't behave as expected.

Current problems: This is written using Base.Cartesian, so it won't work in 0.2. I had a bit of trouble getting the Cartesian.jl version of `@ngenerate` to work. I'll revisit that later (or someone with a bit more Cartesian-fu can take a crack at it :wink:), but until then, I've left it in a separate file rather than merging it with algorithms.jl.

### CODE REVIEW

File: src/bwdist.jl [hunk 1]
```diff
@@ -0,0 +1,156 @@
+using Base.Cartesian
+
+# A function to permute a tuple of subscripts given a
+# permutation vector
+function permutesubs(subs::(Int...),perm::Vector{Int})
```
Comment (juliohm) [inline src/bwdist.jl]: Do we need this function? Why not use `subs[perm]`?

File: src/bwdist.jl [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    s
+end
+
+sqeuclidean(a::AbstractVector, b::AbstractVector) = sumsqdiff(a,b)
+euclidean(a::AbstractVector,b::AbstractVector) = sqrt(sumsqdiff(a,b))
+
+# Be careful with this: you really need an Array of 0s and 1s
+bwdist(I::AbstractArray{Real}) = bwdist(convert(Bool,I))
+# bwdist: call this one!
+@ngenerate N (Array{Int},Array{Int}) function bwdist{N}(I::AbstractArray{Bool,N})
```
Comment (juliohm) [inline src/bwdist.jl]: `@ngenerate` is deprecated in `VERSION >= v"0.4.0-dev+3184"`

Comment (timholy) [inline src/bwdist.jl]: True, but it will be available in `Compat` for some time yet.

**Comment by iamed2** [comment]
Is this a thing that will be merged at some point?

**Comment by timholy** [comment]
It's clearly something of great interest. I've thought several times about merging it, but the fact that it's a lot slower than Matlab still concerns me. Someone will file a performance issue, and then what happens. Do I have to fix it myself? I don't use this functionality (yet), and have far too many packages I maintain.

But I hadn't remembered that the gap was "only" 10x, so maybe it's not so horrible. I was still remembering the 100x days.

Anyone else want to chime in? I could be persuaded either way. But it would need tests and documentation first (if @wkearn doesn't feel like adding them, someone else could fork this branch, add tests, and submit a new PR).

**Comment by rsrock** [comment]
Having some code for this is better than none. That will encourage one of us to write tests, docs, and improve performance. The perfect is the enemy of the good, etc.  I'm in favor of merging.

**Comment by wkearn** [comment]
I'm sorry I've neglected this for so long. I stopped needing a distance transform for my own work, and this fell by the wayside. Let me look at this again this week and see if I can clean it up and improve the performance. Unless someone else has good ideas about how to improve it right now.

**Comment by timholy** [comment]
I've not looked into this algorithm at all, but one comment is that multidimensional stuff is a lot easier now thanks to http://docs.julialang.org/en/latest/manual/arrays/#iteration (you can add/subtract/min/max `CartesianIndex`es) and http://docs.julialang.org/en/latest/manual/metaprogramming/#generated-functions.

**Comment by muhlba91** [comment]
note: work carried on in PR #576

**Comment by timholy** [comment]
Closed by #580

**Pull request closed without merging**